File: 08. One More Time/finitefield/Python/polynomial.py
import argparse
import os

# Modular arithmetic functions for Galois Field
def add(a, b, P):
    return (a + b) % P

def subtract(a, b, P):
    return (a - b) % P

def multiply(a, b, P):
    return (a * b) % P

def inverse(a, P):
    # Find multiplicative inverse using Extended Euclidean Algorithm
    t, newt = 0, 1
    r, newr = P, a

    while newr != 0:
        quotient = r // newr
        t, newt = newt, t - quotient * newt
        r, newr = newr, r - quotient * newr

    if r > 1:
        raise Exception("a is not invertible")
    if t < 0:
        t += P

    return t

def divide(a, b, P):
    return multiply(a, inverse(b, P), P)

def interpolate(d, x, y, P):
    p = [0 for _ in range(d)]

    for i in range(d):
        product = 1
        t = [0 for _ in range(d)]
        for j in range(d):
            if i == j:
                continue
            product = multiply(product, subtract(x[i], x[j], P), P)
        product = divide(y[i], product, P)
        t[0] = product
        for j in range(d):
            if i == j:
                continue
            for k in range(d - 1, 0, -1):
                t[k] = add(t[k], t[k - 1], P)
                t[k - 1] = multiply(t[k - 1], -x[j], P)
        for j in range(d):
            p[j] = add(p[j], t[j], P)

    return p

def main():
    d = 0
    x = []
    y = []

    parser = argparse.ArgumentParser(description="Process the path and prime argument.")
    parser.add_argument("P", type=int, help="Prime number for Galois Field")
    parser.add_argument("path", type=str, help="Path to the file")

    args = parser.parse_args()

    if args.path:
        if not os.path.exists(args.path):
            print(f"Error: The file at '{args.path}' doesn't exist!")
            return
        else:
            with open(args.path, "r") as f:
                try:
                    for line in f:
                        split = line.split(" ")
                        if len(split) != 2:
                            raise ValueError("Invalid number of values in a line")
                        x.append(int(split[0]))
                        y.append(int(split[1]))
                        d += 1
                    p = interpolate(d, x, y, args.P)
                    for i in p:
                        print(i, end=" ")
                    print()
                except:
                    print("Error: Invalid input format")

if __name__ == "__main__":
    main()

File: 08. One More Time/finitefield/Go/polynomial.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func mod(a int, p int) int {
	return (a%p + p) % p
}

func add(a int, b int, p int) int {
	return mod(a+b, p)
}

func sub(a int, b int, p int) int {
	return mod(a-b, p)
}

func mul(a int, b int, p int) int {
	return mod(a*b, p)
}

func inv(a int, p int) int {

	t := 0
	newt := 1
	r := p
	newr := a

	for newr != 0 {
		quotient := r / newr
		temp := t
		t = newt
		newt = temp - quotient*newt
		temp = r
		r = newr
		newr = temp - quotient*newr
	}

	if t < 0 {
		t += p
	}

	return t

}

func div(a int, b int, p int) int {
	return mul(a, inv(b, p), p)
}

func interpolate(d int, P int, x []int, y []int) []int {
	p := make([]int, d)

	for i := 0; i < d; i++ {
		product := 1
		t := make([]int, d)

		for j := 0; j < d; j++ {
			if i == j {
				continue
			}
			product = mul(product, sub(x[i], x[j], P), P)
		}

		product = div(y[i], product, P)
		t[0] = product

		for j := 0; j < d; j++ {
			if i == j {
				continue
			}
			for k := d - 1; k > 0; k-- {
				t[k] = add(t[k-1], t[k], P)
				t[k-1] = mul(t[k-1], -x[j], P)
			}
		}

		for j := 0; j < d; j++ {
			p[j] = add(p[j], t[j], P)
		}
	}

	return p
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println("Error: Please provide a prime number and a file path.")
		return
	}

	P, err3 := strconv.Atoi(os.Args[1])
	if err3 != nil {
		fmt.Println("Error: Not a number")
		return
	}
	path := os.Args[2]
	file, err := os.Open(path)
	if err != nil {
		fmt.Printf("Error: The file at '%s' doesn't exist!\n", path)
		return
	}
	defer file.Close()

	var x, y []int
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		split := strings.Split(line, " ")

		if len(split) != 2 {
			fmt.Println("Error: Invalid input format")
			return
		}

		xVal, err1 := strconv.Atoi(split[0])
		yVal, err2 := strconv.Atoi(split[1])
		if err1 != nil || err2 != nil {
			fmt.Println("Error: Invalid number format in the file")
			return
		}

		x = append(x, xVal)
		y = append(y, yVal)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading the file.")
		return
	}

	p := interpolate(len(x), P, x, y)
	for _, val := range p {
		fmt.Print(val, " ")
	}
}
